src/main/scala/mesosphere/marathon/core/storage/repository/impl/TemplateRepository.scala

```diff
@@ -0,0 +1,174 @@
+  *   implementation for details)</li>
+  * <li> `$safePathId` is [[PathId.safePath]] (see [[mesosphere.marathon.core.storage.repository.impl.TemplateRepository.VersionBucketPath]]
+  *   implementation for details)</li>
+  * <li>`$version` is template version</li>
+  * </ul>
+  *
+  * So two templates with [[PathId]]s `/eng/foo` (with versions 1 and 2) and `/sales/bar` (versions 3) will be stored like following:
+  * {{{
+  * /templates
+  *   /01
+  *     /eng_foo
+  *       /1
+  *       /2
+  *   /05
+  *     /sales_bar
+  *       /3
+  * }}}
+  * Most methods simply use the underlying [[BucketingRepository]] methods to store templates by mapping from [[PathId]] and version
+  * to [[BucketNode]] and [[BucketPath]] using [[mesosphere.marathon.core.storage.repository.impl.TemplateRepository.VersionBucketPath]]
+  * and [[mesosphere.marathon.core.storage.repository.impl.TemplateRepository.TemplateBucketPath]] helper classes.
+  *
+  * [[create]] method takes a [[Spec]] and returns a [[Template]] which contains the spec and it's version. A version is an integer that
+  * is automatically generated and assigned by the repository. Internally this repository holds a counter for every template incrementing
+  * it every time a new template version is created.
+  *
+  * [[update()]] method is overridden and throws an [[UnsupportedOperationException]] since templates are immutable and should not
+  * be updated.
+  *
+  * [[init()]] method initializes repository by reading all existing templates Ids and their versions (but not template content) and
+  * initializing internal template version counters. This method should be called and waited for successful competition prior to any
+  * other usage of the repository.
+  *
+  * Known pitfalls: please note that while there is one at least one version of the template, a counter is always increased monotonically.
+  * However should all template versions be deleted and marathon restarted, a counter is initialized with `0`.
+  * This implementation is tread-safe and provides serializability guarantees e.g. two independent [[create()]] calls are not guaranteed
+  * to be executed in any particular order the same way one should not expect two independent futures to have an order.
+  *
+  * @param store underlying instance of [[ZooKeeperPersistenceStore]]
+  * @param ec execution context
+  * @param mat stream materializer
+  */
+class TemplateRepository(val store: ZooKeeperPersistenceStore)(implicit ec: ExecutionContext, mat: ActorMaterializer)
+  extends BucketingRepository with TemplateRepositoryLike with StrictLogging {
+
+  import TemplateRepository._
+
+  override val base: String = "/templates"
+
+  /**
+    * We store a counter for the template versions separately. It's helpful when calculating next template
+    * version. A counter is initialized by the current template max version number in the [[init()]] method.
+    */
+  val counters = new ConcurrentHashMap[PathId, AtomicInteger]()
```

Comment:
Why not set the counter after `init`? There current implementations hides accesses to `counters` before it was initialized. `var counters = _` seems ugly but would surface an init issue with a `NullPointerException`. Also, why is the value an `AtomicInteger` if we use `compute`?